File: src/main/java/dev/sky_lock/pocketlifevehicle/vehicle/Engine.java
package dev.sky_lock.pocketlifevehicle.vehicle;

import dev.sky_lock.pocketlifevehicle.vehicle.model.MaxSpeed;
import dev.sky_lock.pocketlifevehicle.vehicle.model.Model;

/**
 * @author sky_lock
 */

class Engine {
    private final CarStatus status;
    private final Model model;
    private final Speed speed;
    private float currentSpeed;

    Engine(CarStatus status, Model model) {
        this.status = status;
        this.model = model;
        this.speed = status.getSpeed();
    }

    void consumeFuel(float sideInput) {
        if (speed.approximate() == 0.0F) {
            if (sideInput == 0.0F) {
                return;
            }
            if (status.getFuel() - 0.05F <= 0.0F) {
                status.setFuel(0.0F);
                return;
            }
        }
        status.setFuel(status.getFuel() - 0.05F);
    }

    boolean refuel(float added) {
        float current = status.getFuel();
        float max = model.getMaxFuel();
        if (current >= max) {
            return false;
        }
        if (current + added > max) {
            status.setFuel(max);
            return true;
        }
        status.setFuel(current + added);
        return true;
    }


    void update(float passengerInput) {
        if (passengerInput == 0.0f) {
            if (speed.isPositive()) {
                speed.frictionalDecelerate();
            }
        } else if (passengerInput < 0.0f) {
            speed.decelerate();
        }

        MaxSpeed maxSpeed;
        if (model.getMaxSpeed() == null) {
            maxSpeed = MaxSpeed.NORMAL;
        } else {
            maxSpeed = model.getMaxSpeed();
        }

        if (speed.exact() > maxSpeed.getMax()) {
            if (status.getFuel() <= 0.0f) {
                speed.zeroize();
            }
            this.currentSpeed = speed.exact();
            return;
        }
        if (passengerInput > 0.0f) {
            speed.accelerate();
        }
        if (speed.isNegative()) {
            speed.decrease();
        }
        if (status.getFuel() <= 0.0f) {
            speed.zeroize();
        }
        this.currentSpeed = speed.exact();
    }

    float getCurrentSpeed() {
        return currentSpeed;
    }

    void stop() {
        status.getSpeed().zeroize();
    }

    float speedPerSecond() {
        return speed.exact() * 20;
    }

}

File: src/main/kotlin/dev/sky_lock/pocketlifevehicle/vehicle/Engine.kt
package dev.sky_lock.pocketlifevehicle.vehicle

import dev.sky_lock.pocketlifevehicle.vehicle.model.Model

/**
 * @author sky_lock
 */
private const val ZERO = 0.0F

class Engine(private val tank: FuelTank, private val model: Model) {
    val speed = Speed()
    var currentSpeed = ZERO
        private set

    fun update(sideIn: Float, forIn: Float) {
        if (tank.isEmpty()) {
            speed.zeroize()
            currentSpeed = speed.exact()
            return
        }
        if (speed.exact() > model.spec.maxSpeed.value) {
            if (forIn < ZERO) {
                speed.decelerate()
            } else if (forIn == ZERO) {
                speed.frictionalDecelerate()
            }
            currentSpeed = speed.exact()
            if (!model.flag.eventOnly && model.flag.consumeFuel) {
                tank.consume()
            }
            return
        }
        if (forIn == ZERO) {
            if (speed.isPositive) {
                speed.frictionalDecelerate()
            }
        } else if (forIn < ZERO) {
            speed.decelerate()
        } else {
            speed.accelerate()
        }
        if (speed.isNegative) {
            speed.decrease()
        }
        currentSpeed = speed.exact()
        if (speed.approximate() == ZERO && sideIn == ZERO) {
            return
        }
        if (!model.flag.eventOnly && model.flag.consumeFuel) {
            tank.consume()
        }
    }

    fun stop() {
        speed.zeroize()
    }

    fun speedPerSecond(): Float {
        return speed.exact() * 20
    }
}